rust/rope/src/diff.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Computing deltas between two ropes.
+
+use std::collections::HashMap;
+
+use compare::RopeScanner;
+use delta::{Delta, DeltaElement};
+use interval::Interval;
+use rope::{LinesMetric, Rope, RopeInfo, RopeDelta};
+use tree::{Node, NodeInfo};
+
+use memchr::memchr;
+
+/// A trait implemented by various diffing strategies.
+pub trait Diff<N: NodeInfo> {
+    fn compute_delta(base: &Node<N>, target: &Node<N>) -> Delta<N>;
+}
+
+/// The minimum length of non-whitespace characters in a line before
+/// we consider it for diffing purposes.
+const MIN_SIZE: usize = 32;
+
+/// A line-oriented, hash based diff algorithm.
+///
+/// This works by taking a hash of each line in either document that
+/// has a length, ignoring leading whitespace, above some threshold.
+///
+/// Lines in the target document are matched against lines in the
+/// base document. When a match is found, it is extended forwards
+/// and backwards as far as possible.
+///
+/// This runs in O(n+m) in the lengths of the two ropes, and produces
+/// results on a variety of workloads that are comparable in quality
+/// (measured in terms of serialized diff size) with the results from
+/// using a suffix array, while being an order of magnitude faster.
+pub struct LineHashDiff;
+
+impl Diff<RopeInfo> for LineHashDiff {
+    fn compute_delta(base: &Rope, target: &Rope) -> RopeDelta {
+        let mut builder = DiffBuilder::default();
+
+        // before doing anything, scan top down and bottom up for like-ness.
+        let mut scanner = RopeScanner::new(base, target);
+        let (start_offset, diff_end) = scanner.find_min_diff_range();
+        let target_end = target.len() - diff_end;
+
+        if start_offset > 0 {
+            builder.copy(0, 0, start_offset);
+        }
+
+        // if our preliminary scan finds no differences we're done
+        if start_offset == base.len() && target.len() == base.len() {
+            return builder.to_delta(base, target);
+        }
+
```

Comment:
As discussed in irc, here's my proposal for the way forward, which should be a follow-on issue.

Create a RopeSlice struct, which is a pair of (&Rope, Range). Custom implement Eq and Hash traits, to internally use iter_chunks to compute both. Don't use lines_raw iterator, use cursor with LinesMetric. I think this gives you maximum efficiency, as everything is borrowed and there's no allocation or copying, even in the case of very long lines.

I did verify that you can do DefaultHasher::write() with different slice boundaries and get the same state, though I don't believe that's written down explicitly.